### DISCUSSION TITLE
Sorting activities within invoices (per-project)

### DISCUSSION THREAD

**Discussion by Cayburn** [kimai/kimai#3477] [Questions & Answers] [answered]
Is there a sorting opportunity for invoices? We are using per-project invoices and would like the entries to be sorted - ideally by a meta-field or alternatively by the activity name. We are using DOCX templates. I could imagine on twig it's possible as there are more programming commands possible, but ideally we prefere to stay with DOCX.

**Comment by kevinpapst** [comment, marked as answer]
Add you own invoice calculator:
https://github.com/kevinpapst/kimai2/blob/master/src/Invoice/Calculator/ActivityInvoiceCalculator.php

I guess you can overwrite `public function getEntries(): array` and add your sorting logic.

**Reply by Cayburn** [reply]
Thanks! I will do this :)

**Reply by Cayburn** [reply]
Thanks, the following function is doing the trick. Do you probably have a hint on how to configure that my modified ActivityInvoiceCalculator is used instead of the default one? Overwriting might not be the best option. 

        $result_array = array_values($invoiceItems);
        usort($result_array, function ($a, $b) {
            $compare_a = $a->getActivity()->getName();
            $compare_b = $b->getActivity()->getName();

            if ($compare_a == $compare_b) {
                return 0;
            };
            return ($compare_a < $compare_b) ? -1 : 1;
        });

        return $result_array;

**Reply by kevinpapst** [reply]
Yours should show up as new one in the invoice template dropdown.
What is the issue with having a new calculator? 
Give it a clear name (you can translate the ID through the file https://github.com/kevinpapst/kimai2/blob/master/translations/invoice-calculator.de.xlf which you can add to your customer specific plugin) and tell your customer not to change it.

**Reply by Cayburn** [reply]
I have the feeling I am always one Step away. Using a custom name calculator in source is now working (forgot to use a unique getId). Now I tried to bring it to a Bundle, so updating the DemoBundle. Do I need to register something somewhere? The RandomNumberGenerator seems to work without further references.

https://github.com/KatjaGlassConsulting/DemoBundle

Thanks for your hints.

**Reply by kevinpapst** [reply]
Looks ok to me, but haven't tested. Did you try to `rm -rf var/cache/*`

**Reply by Cayburn** [reply]
I tried with no success. Did the usual clear chache, reload and now also tried with manual removed cache.
